This seems to have uncovered a flaky behavior in the connection pool code on ARM64:

```
=== RUN   TestStressCloseDuringTraffic/cycle-054
2026-06-10 09:46:01.128 UTC ERR smartconnpool/pool.go:375 failed to reopen pool "": Code: ABORTED
timed out while waiting for connections to be returned to the pool (capacity=0, active=6, borrowed=0)

    stress_test.go:1016: 
        	Error Trace:	/home/ubuntu/_work/vitess/vitess/go/pools/smartconnpool/stress_test.go:1183
        	            				/home/ubuntu/_work/vitess/vitess/go/pools/smartconnpool/stress_test.go:1016
        	Error:      	Condition never satisfied
        	Test:       	TestStressCloseDuringTraffic/cycle-054
        	Messages:   	cycle 54: CloseWithContext stalled: capacity=0 active=10 borrowed=0 open=144 isOpen=true liveGetWorkers=24 capacityInProgress=true connectsAfterClose=0
    stress_test.go:1016: 
        	Error Trace:	/home/ubuntu/_work/vitess/vitess/go/pools/smartconnpool/stress_test.go:1193
        	            				/home/ubuntu/_work/vitess/vitess/go/pools/smartconnpool/stress_test.go:1016
        	Error:      	CloseWithContext stalled
        	Test:       	TestStressCloseDuringTraffic/cycle-054
        	Messages:   	cycle 54: capacity=0 active=6 borrowed=0 open=149 isOpen=false liveGetWorkers=0 capacityInProgress=false connectsAfterClose=5
--- FAIL: TestStressCloseDuringTraffic/cycle-054 (30.01s)
```

--- a/.github/actions/setup-mysql/action.yml
+++ b/.github/actions/setup-mysql/action.yml
@@ -22,40 +22,51 @@ runs:
         sudo rm -rf /var/lib/mysql
         sudo rm -rf /etc/mysql
 
-        # We have to install this old version of libaio1. See also:
-        # https://bugs.launchpad.net/ubuntu/+source/libaio/+bug/2067501
-        wget http://archive.ubuntu.com/ubuntu/pool/main/liba/libaio/libaio1_0.3.112-13build1_amd64.deb && \
-          sudo dpkg -i libaio1_0.3.112-13build1_amd64.deb && \
-          rm libaio1_0.3.112-13build1_amd64.deb
-
-        # Get key to latest MySQL repo
-        sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys A8D3785C
-        wget -c https://dev.mysql.com/get/mysql-apt-config_0.8.35-1_all.deb
-
-        if [[ "${{ inputs.flavor }}" == "mysql-5.7" ]]; then
-          # Bionic packages are still compatible for Jammy since there's no MySQL 5.7
-          # packages for Jammy.
-          echo mysql-apt-config mysql-apt-config/repo-codename select bionic | sudo debconf-set-selections
-          echo mysql-apt-config mysql-apt-config/select-server select mysql-5.7 | sudo debconf-set-selections
-          sudo DEBIAN_FRONTEND="noninteractive" dpkg -i mysql-apt-config*
-          sudo apt-get update
-          # libtinfo5 is also needed for older MySQL 5.7 builds.
-          curl -L -O http://archive.ubuntu.com/ubuntu/pool/universe/n/ncurses/libtinfo5_6.3-2ubuntu0.1_amd64.deb
-          sudo dpkg -i libtinfo5_6.3-2ubuntu0.1_amd64.deb
-          sudo DEBIAN_FRONTEND="noninteractive" apt-get install -y mysql-client=5.7* mysql-community-server=5.7* mysql-server=5.7* libncurses6
-        elif [[ "${{ inputs.flavor }}" == "mysql-8.0" ]]; then
-          echo mysql-apt-config mysql-apt-config/select-server select mysql-8.0 | sudo debconf-set-selections
-          sudo DEBIAN_FRONTEND="noninteractive" dpkg -i mysql-apt-config*
-          sudo apt-get update
-          sudo DEBIAN_FRONTEND="noninteractive" apt-get install -y mysql-server mysql-client
-        elif [[ "${{ inputs.flavor }}" == "mysql-8.4" ]]; then
-          echo mysql-apt-config mysql-apt-config/select-server select mysql-8.4-lts | sudo debconf-set-selections
-          sudo DEBIAN_FRONTEND="noninteractive" dpkg -i mysql-apt-config*
-          sudo apt-get update
+        if [[ "$(dpkg --print-architecture)" == "arm64" ]]; then
+          # Oracle's apt repo only publishes amd64 packages for Ubuntu, so on
+          # arm64 we install MySQL from the Ubuntu archive instead, which only
+          # ships MySQL 8.0.
+          if [[ "${{ inputs.flavor }}" != "mysql-8.0" ]]; then
+            echo "Unsupported MySQL flavor on arm64: ${{ inputs.flavor }}"
+            exit 1
+          fi
           sudo DEBIAN_FRONTEND="noninteractive" apt-get install -y mysql-server mysql-client
         else
-          echo "Unsupported MySQL flavor: ${{ inputs.flavor }}"
-          exit 1
+          # We have to install this old version of libaio1. See also:
+          # https://bugs.launchpad.net/ubuntu/+source/libaio/+bug/2067501
+          wget http://archive.ubuntu.com/ubuntu/pool/main/liba/libaio/libaio1_0.3.112-13build1_amd64.deb && \
+            sudo dpkg -i libaio1_0.3.112-13build1_amd64.deb && \
+            rm libaio1_0.3.112-13build1_amd64.deb
+
+          # Get key to latest MySQL repo
+          sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys A8D3785C
+          wget -c https://dev.mysql.com/get/mysql-apt-config_0.8.35-1_all.deb
+
+          if [[ "${{ inputs.flavor }}" == "mysql-5.7" ]]; then
+            # Bionic packages are still compatible for Jammy since there's no MySQL 5.7
+            # packages for Jammy.
+            echo mysql-apt-config mysql-apt-config/repo-codename select bionic | sudo debconf-set-selections
+            echo mysql-apt-config mysql-apt-config/select-server select mysql-5.7 | sudo debconf-set-selections
+            sudo DEBIAN_FRONTEND="noninteractive" dpkg -i mysql-apt-config*
+            sudo apt-get update
+            # libtinfo5 is also needed for older MySQL 5.7 builds.
+            curl -L -O http://archive.ubuntu.com/ubuntu/pool/universe/n/ncurses/libtinfo5_6.3-2ubuntu0.1_amd64.deb
+            sudo dpkg -i libtinfo5_6.3-2ubuntu0.1_amd64.deb
+            sudo DEBIAN_FRONTEND="noninteractive" apt-get install -y mysql-client=5.7* mysql-community-server=5.7* mysql-server=5.7* libncurses6
+          elif [[ "${{ inputs.flavor }}" == "mysql-8.0" ]]; then
+            echo mysql-apt-config mysql-apt-config/select-server select mysql-8.0 | sudo debconf-set-selections
+            sudo DEBIAN_FRONTEND="noninteractive" dpkg -i mysql-apt-config*
+            sudo apt-get update
+            sudo DEBIAN_FRONTEND="noninteractive" apt-get install -y mysql-server mysql-client
+          elif [[ "${{ inputs.flavor }}" == "mysql-8.4" ]]; then
+            echo mysql-apt-config mysql-apt-config/select-server select mysql-8.4-lts | sudo debconf-set-selections
+            sudo DEBIAN_FRONTEND="noninteractive" dpkg -i mysql-apt-config*
+            sudo apt-get update
+            sudo DEBIAN_FRONTEND="noninteractive" apt-get install -y mysql-server mysql-client
+          else
+            echo "Unsupported MySQL flavor: ${{ inputs.flavor }}"
+            exit 1
+          fi
         fi
 
         sudo service mysql stop

--- a/.github/workflows/unit_test.yml
+++ b/.github/workflows/unit_test.yml
@@ -57,6 +57,18 @@ jobs:
             evalengine: "1"
             flavor: mysql-8.4
 
+          # arm64 tests (mysql80 only — Oracle's apt repo has no arm64
+          # packages and Ubuntu's archive only ships MySQL 8.0 on arm64)
+          - platform: mysql80
+            evalengine: "0"
+            flavor: mysql-8.0
+            arch: arm64
+
+          - platform: mysql80
+            evalengine: "1"
+            flavor: mysql-8.0
+            arch: arm64
+
           # Race tests (mysql80 only)
           - platform: mysql80
             evalengine: "0"
@@ -69,8 +81,8 @@ jobs:
             flavor: mysql-8.0
             race: true
 
-    name: "Unit Test (${{ matrix.race && (matrix.evalengine == '1' && 'Evalengine_' || '') || (matrix.evalengine == '1' && 'evalengine_' || '') }}${{ matrix.race && 'Race' || matrix.platform }})"
-    runs-on: ${{ matrix.race && github.repository == 'vitessio/vitess' && 'oracle-vm-16cpu-64gb-x86-64' || 'ubuntu-24.04' }}
+    name: "Unit Test (${{ matrix.race && (matrix.evalengine == '1' && 'Evalengine_' || '') || (matrix.evalengine == '1' && 'evalengine_' || '') }}${{ matrix.race && 'Race' || matrix.platform }}${{ matrix.arch && format('_{0}', matrix.arch) || '' }})"
+    runs-on: ${{ github.repository == 'vitessio/vitess' && (matrix.arch == 'arm64' && 'oracle-vm-16cpu-64gb-arm64' || 'oracle-vm-16cpu-64gb-x86-64') || (matrix.arch == 'arm64' && 'ubuntu-24.04-arm' || 'ubuntu-24.04') }}
 
     steps:
       - name: Harden the runner (Audit all outbound calls)

